src/clang-c-frontend/clang_c_convert.cpp

```diff
@@ -1444,7 +1446,54 @@ bool clang_c_convertert::get_expr(const clang::Stmt &stmt, exprt &new_expr)
     if(get_expr(*compound.getInitializer(), initializer))
       return true;
 
-    new_expr = initializer;
+    typet t = initializer.type();
+
+    /* A compound literal is an LValue that has associated static or automatic
+     * storage, depending on whether it appears in file or block scope.
+     * Therefore, in C, for instance a pointer to it can be taken and used
+     * within the same block without restrictions, e.g.
+     *
+     *   int *p = &(int){0}, x = 42;
+     *   memcpy(p, &x, sizeof(int));
+     *   assert(*p == 42);
+     *
+     * It has the same semantics as an non-anonymous variable of the
+     * corresponding type. Thus, we introduce a new symbol to represent it.
+     */
+
+    /* "cl" stands for "compound literal", but that is too long for possibly
+     * frequent copies of the symbol's name and id. */
+    std::string module_name =
+      get_modulename_from_path(location.file().as_string());
+    symbolt &cl = anon_symbol.new_symbol(context, t, module_name + "$cl$");
```

Comment:
Wouldn't be better to be verbose here? I mean, I feel that in the CE the user will see this `$cl$` and will be lost.

Reply:
You are right.
```
State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main$cl$1.raw
```
I now changed it to:
```
State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main.c:14$compound-literal$1.raw
```

Reply:
we can also mark this assignment as hidden. but changing symex_assign, that way it won't appear on the cex unless we force printing the assignment.

Do we print this variable when creating a witness? It shouldn't be there.

Reply:
> we can also mark this assignment as hidden. but changing symex_assign, that way it won't appear on the cex unless we force printing the assignment.

Sorry, I don't understand, can you elaborate? Which assignment are you referring to (`var = ...` or the initialization of the compound-literal symbol?) and how can printing the assignment be forced?

Reply:
> Do we print this variable when creating a witness? It shouldn't be there.

This is the entire cex output for https://github.com/esbmc/esbmc/blob/b3ad1994b3d2f3cfb1dd9847b08a389368372703/regression/esbmc/github_709-cl/main.c
```
Counterexample:

State 1 file main.c line 14 function main thread 0
----------------------------------------------------
  var = main.c:14$compound-literal$1.raw

State 2 file main.c line 16 function main thread 0
----------------------------------------------------
  ta = 127 (00000000 00000000 00000000 01111111)

State 3 file main.c line 19 function main thread 0
----------------------------------------------------
Violated property:
  file main.c line 19 function main
  assertion ta == tb
  ta == tb


VERIFICATION FAILED
```

Reply:
@mikhailramalho meant when you use the `witness-output` option. Witness validators will not be able to handle the compound literal assignment

Reply:
@rafaelsamenezes Thanks, I wasn't aware.

@mikhailramalho this is the [witness](https://github.com/esbmc/esbmc/files/8765153/witness.txt) output. It does not contain the identifier nor the assignment (only `ta = 127;`), however, it does show
```xml
    <edge id="E0" source="N0" target="N1">
      <data key="enterFunction">main</data>
      <data key="createThread">0</data>
    </edge>
```
as the first edge in the graph corresponding to the assignment.

I have zero experience with these witness files; is it ok for this PR?

Reply:
Actually, the witness looks fine.

May I ask if you can do one last test? Can you initialize .raw and see what we get in the cex/witness?

Reply:
`int var = (s2){.raw = (&x)->data}.raw;` makes no difference.

`int var = ((s2){(&x)->data}.raw = 42);` obviously changes `ta = 42`, but that's it: [cex](https://github.com/esbmc/esbmc/files/8770055/cex.txt), [witness.txt](https://github.com/esbmc/esbmc/files/8770059/witness.txt)

`int var = ((s2){(&x)->data}.raw = (&x)->data);` again no change to the previous.

Reply:
If `var` becomes a struct type, the concrete value is printed, but the compound literal identifier is not: 
[test.c](https://github.com/esbmc/esbmc/files/8770086/test.c.txt), [cex.txt](https://github.com/esbmc/esbmc/files/8770090/cex.txt), [witness.txt](https://github.com/esbmc/esbmc/files/8770102/witness.txt)

Reply:
nice, so now my point is if we should print the:
```
State 1 file test.c line 14 function main thread 0
----------------------------------------------------
  var = test.c:14$compound-literal$1.raw
```

Can you hide it from the cex? I mean, unless there is a value, we don't usually print these assignments. WDYT @lucasccordeiro, @rafaelsamenezes?

Reply:
I find it rather confusing that the value of `var` is not given in the counter example. For me it would make more sense if the line read
```
  var = { .raw=127 }.raw
```
I mean, consider this shortened [example](https://github.com/esbmc/esbmc/files/8770278/test.c.txt)
```c
  int var = (s2){(&x)->data}.raw;
  int ta;
  ta = var;
  var = 0;
  assert(ta == var);
```
with nearly incomprehensible counter-example since `var`'s value is not shown:
```
Counterexample:

State 1 file test.c line 14 function main thread 0
----------------------------------------------------
  var = test.c:14$compound-literal$1.raw

State 2 file test.c line 16 function main thread 0
----------------------------------------------------
  ta = 127 (00000000 00000000 00000000 01111111)

State 3 file test.c line 18 function main thread 0
----------------------------------------------------
Violated property:
  file test.c line 18 function main
  assertion ta == var
  ta == var
```